src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -0,0 +1,712 @@
+
+using namespace leader_watcher;
+
+using librbd::util::create_context_callback;
+using librbd::util::create_rados_ack_callback;
+
+LeaderWatcher::LeaderWatcher(Threads *threads, librados::IoCtx &io_ctx)
+  : Watcher(io_ctx, threads->work_queue, RBD_MIRROR_LEADER),
+    m_threads(threads),
+    m_lock("rbd::mirror::LeaderWatcher " + io_ctx.get_pool_name()),
+    m_notifier_id(librados::Rados(io_ctx).get_instance_id()) {
+}
+
+int LeaderWatcher::init() {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(!m_leader_lock);
+
+  int r = m_ioctx.create(m_oid, false);
+  if (r < 0) {
+    derr << "error creating " << m_oid << " object: " << cpp_strerror(r)
+         << dendl;
+    return r;
+  }
+
+  m_leader_lock.reset(
+    new LeaderLock(m_ioctx, m_work_queue, m_oid, this, true,
+                   g_ceph_context->_conf->rbd_blacklist_expire_seconds));
+
+  C_SaferCond register_ctx;
+  register_watch(&register_ctx);
+  r = register_ctx.wait();
+  if (r < 0) {
+    derr << "error registering leader watcher for " << m_oid << " object: "
+         << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  acquire_leader_lock();
+  return 0;
+}
+
+void LeaderWatcher::shut_down() {
+  dout(20) << dendl;
+
+  Mutex::Locker timer_locker(m_threads->timer_lock);
+  Mutex::Locker locker(m_lock);
+
+  while (true) {
+    assert(m_leader_lock);
+
+    if (m_leader_lock->is_lock_owner()) {
+      C_SaferCond release_lock_ctx;
+      m_leader_lock->release_lock(async(&release_lock_ctx));
+      m_lock.Unlock();
+      m_threads->timer_lock.Unlock();
+      int r = release_lock_ctx.wait();
+      if (r < 0) {
+	derr << "error releasing leader lock: " << cpp_strerror(r) << dendl;
+      }
+      m_threads->timer_lock.Lock();
+      m_lock.Lock();
+      continue;
+    }
+
+    cancel_timer_task();
+    break;
+  }
+
+  assert(m_leader_lock);
+  C_SaferCond shutdown_lock_ctx;
+  m_leader_lock->shut_down(&shutdown_lock_ctx);
+  m_lock.Unlock();
+  m_threads->timer_lock.Unlock();
+  int r = shutdown_lock_ctx.wait();
+  if (r < 0) {
+    derr << "error shutting down leader lock: " << cpp_strerror(r) << dendl;
+  }
+
+  C_SaferCond unregister_ctx;
+  unregister_watch(&unregister_ctx);
+  r = unregister_ctx.wait();
+  if (r < 0) {
+    derr << "error unregistering leader watcher for " << m_oid << " object: "
+	 << cpp_strerror(r) << dendl;
+  }
+
+  m_threads->timer_lock.Lock();
+  m_lock.Lock();
+  m_leader_lock.reset();
+}
+
+bool LeaderWatcher::is_leader() {
+  Mutex::Locker locker(m_lock);
+
+  return is_leader(m_lock);
+}
+
+bool LeaderWatcher::is_leader(Mutex &lock) {
+  assert(m_lock.is_locked());
+
+  bool leader = m_leader_lock && m_leader_lock->is_leader();
+  dout(20) << leader << dendl;
+  return leader;
+}
+
+void LeaderWatcher::release_leader() {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  if (!is_leader(m_lock)) {
+    return;
+  }
+
+  release_leader_lock();
+}
+
+Context *LeaderWatcher::async(Context * ctx) {
```

Comment:
Nit: rename to ```create_async_context```? or update the existing ```librbd::util::create_async_context``` to be able to handle either an ```ImageCtx```-like param or a ```ContextWQ```-like param.